perf: Cache db columns to avoid redundant database calls

surajshetty3416 opened the pull request:
While saving the user form:

**Redundant calls**
<img width="1093" alt="Screenshot 2019-10-02 at 7 47 46 PM" src="https://user-images.githubusercontent.com/13928957/66053633-4b00d180-e550-11e9-9573-1905d0e9e81d.png">
<img width="1089" alt="Screenshot 2019-10-02 at 7 48 00 PM" src="https://user-images.githubusercontent.com/13928957/66053723-75528f00-e550-11e9-8bc9-4a292b421842.png">


---- 
This should save some `ms` for get queries because it uses `get_table_columns` to validate columns
<img width="1042" alt="Screenshot 2019-10-02 at 8 11 22 PM" src="https://user-images.githubusercontent.com/13928957/66054265-586a8b80-e551-11e9-976d-aac55fbea503.png">

surajshetty3416 committed:
perf: Cache db columns to avoid redundant database calls

hrwX commented:
@surajshetty3416 The `db_colums` change when you add custom fields, shouldn't the cache be clear after adding custom field ?

surajshetty3416 commented:
@hrwX Custom fields are added through Customize Form... and I have cleared `db_column` cache on any update through Customize Form.

**Edit:** I can move the clear cache code from **DocType** and **Customize Form** to `db_table.sync()` code where actual table columns get updated which will make more sense... I'll do that.

surajshetty3416 committed:
fix: Move cache clearing code from doctype to schema.py

surajshetty3416 committed:
fix: self.table_name instead of self.name

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into optimize-db-calls

netchampfaris committed:
fix: Cache columns in  "table_columns" key

`table_columns` was cached in meta but columns were also getting accessed
directly using frappe.db.get_table_columns. Now, it is cached at
`frappe.db` layer

Co-authored-by: Suraj Shetty <surajshetty3416@gmail.com>
